swarm/metrics.py

```diff
@@ -0,0 +1,165 @@
+    for a particular epoch.  Hence, it summarises the predictions made across different bees
+    in a swarm.
+
+    2: This function returns a summary statistic regarding the losses of different bees at each epoch.
+
+    3. This function returns a summary statistic regarding of another summary statistic recorded at each
+    point in the domain. In this case it should be used in with argument `across` set to `x`.
+
+    Note: Could use a refactor to generalise so that any one-d function can be passed in
+    as an argument. (didn't do this initially because the range concept was important, and
+    I wanted this to be accessible along the same lines as mean, max, min etc.)
+
+    Args:
+        data: In use case 1 this should be a numpy array of predictions, as yeilded by core.swarm_train.
+            It will have three dimensions, of structure [bee, epoch/batch, x_step].
+            In use case 2 this should be a numpy array of losses, which can also be yeilded by core.swarm_train
+            It will have two dimensions, of structure [bee, epoch].
+            In use case 3 this should be a numpy array of summary statistics across bees at all points in domain x.
+            It will have two dimensions, of structure [epoch/batch, x_step]
+        func_string: A string indicating the type of summary statistic desired.
+            Currently allowable includes 'min', 'max', 'mean', 'median', 'std', and 'range'
+        across: Takes ony 'Indicate whether you'd like to summarise across the bees (as in use case 1 and 2),
+            or summarise across the a value for each point in domain x.
+
+    Returns: a numpy array of two dimensions, of structure [epoch/batch, x_step] that gives the
+        summary evaluation for each point in x space for all the bees at that epoch of training.
+
+    """
+    valid_axes = {'bees', 'x'}
+    if across not in valid_axes:
+        raise ValueError("argument `across` must be one of %r." % valid_axes)
+
+    # Create a simple dict of likely functions
+    funcs = {
+        "mean": np.mean,
+        "std": np.std,
+        "median": np.median,
+        "min": np.min,
+        "max": np.max
+    }
+
+    # Identify the suitable axis to summarise across
+    if across == "bees":
+        axis_choice = 0
+    elif across == "x":
+        axis_choice = len(data.shape) - 1
+
+    # Create summary of data
+    if func_string in funcs.keys():
+        return np.apply_along_axis(funcs[func_string], axis_choice, data)
+    elif func_string == 'range':
+        point_maxs = np.apply_along_axis(np.max, axis_choice, data)
+        point_mins = np.apply_along_axis(np.min, axis_choice, data)
+        return point_maxs - point_mins
+    else:
+        raise ValueError("Function not in currently allowable of min', 'max', 'mean', 'median', 'std', and 'range'")
+
+
+def rms_error(swarm_ypreds: np.ndarray, y: np.ndarray) -> np.ndarray:
```

Comment:
You should call this `mseloss` to be a bit closer to torch. 

I think this could be useful as a way of taking a loss function and applying it to predictions and handling/documenting the change in dimensionality.

Reply:
Yep, name change a good idea. I briefly tried calling the torch loss function, but it didn't work quickly easily, whilst handling dimensions etc.

Reply:
I think having a function that applies a torch loss to swarm_ypreds could be useful. However I think we can leave it till we really need it.

Reply:
I went for `mse_loss`. I don't much like using cases in function names, and not distinguishing different parts of a string as being separate I also don't like.